chainer/distributions/bernoulli.py

```diff
@@ -59,27 +56,40 @@ def __init__(self, logx):
 
     def forward(self, inputs):
         x, = inputs
-        xp = cuda.get_array_module(x)
-        if x.ndim == 0:
-            x = xp.asarray(x).reshape(1)
-        self.invalid = (x == 0)
+        self.x_zero = (x == 0)
         y = x * self._logx.array
-        y[self.invalid] = 0.
+        y[self.x_zero] = 0.
         return y,
 
     def backward(self, indexes, grad_outputs):
+        if self.x_zero.any():
+            warnings.warn(
+                'cannot calculate gradient for zero input.',
+                RuntimeWarning)
         gy, = grad_outputs
-        dx = (1 + self._logx) * (1 - self.invalid.reshape(self._logx.shape))
+        dx = (1 + self._logx) * (1 - self.x_zero)
         return gy * dx,
 
 
 def _bernoulli_log_prob(logit, x):
+    if x.ndim == 0:
+        x = reshape.reshape(x, (1,))
+    if logit.ndim == 0:
+        logit = reshape.reshape(logit, (1,))
     y, = BernoulliLogProb().apply((logit, x))
     return y
 
 
 def _modified_xlogx(x):
-    return ModifiedXLogX(exponential.log(x)).apply((x,))[0]
+    xp = cuda.get_array_module(x)
+    if x.ndim == 0:
+        x = reshape.reshape(x, (1,))
+    if isinstance(x, chainer.Variable):
+        return ModifiedXLogX(exponential.log(
```

Comment:
Do not repeat the application of `ModifiedXLogX`. How about storing logx, or
```
x = chainer.as_variable(x)
xp = x.xp
```

Reply:
Oh yes!